commerce/domain/tax/cart/calculator/classes/TaxCartCalculatorSample.cls

```diff
@@ -1,349 +1,249 @@
-
-  // You MUST change the useExternalService to True if you want to use the Third Party Service.
-  private static Boolean useExternalService = false;
-  public virtual override void calculate(CartExtension.CartCalculateCalculatorRequest request) {
-    try {
-      CartExtension.Cart cart = request.getCart();
-
-      // Clean up CVO based on tax. When new tax calculator request comes, we need to clean up
-      // previous CVOs as they have been previously handled by the Cart Calculate API.
-      CartExtension.CartValidationOutputList cartValidationOutputCollection = cart.getCartValidationOutputs();
-      Iterator<CartExtension.CartValidationOutput> cartValidationOutputCollectionIterator = cartValidationOutputCollection.iterator();
-      while (cartValidationOutputCollectionIterator.hasNext()) {
-        CartExtension.CartValidationOutput cvo = cartValidationOutputCollectionIterator.next();
-        if (cvo.getType() == CartExtension.CartValidationOutputTypeEnum.TAXES) {
-          cartValidationOutputCollection.remove(cvo);
-        }
-      }
+	// You MUST change this to be your service or you must launch your own Third Party Service and add the host in Setup | Security | Remote site settings.
+	private static String externalTaxHost = 'https://example.com';
+
+	// You MUST change the useExternalService to True if you want to use the Third Party Service.
+	private static Boolean useExternalService = false;
+
+	public virtual override void calculate(CartExtension.CartCalculateCalculatorRequest request) {
+		CartExtension.Cart cart = request.getCart();
+		CartExtension.CartValidationOutputList cartValidationOutputCollection = cart.getCartValidationOutputs();
+		try{
+			Integer cartItemIdSeq = 0;
+			if (!isValidCart(cart)){
+                CartExtension.CartValidationOutput cvo = new CartExtension.CartValidationOutput(
+                CartExtension.CartValidationOutputTypeEnum.TAXES,
+                CartExtension.CartValidationOutputLevelEnum.ERROR);
+                cvo.setMessage('There was a problem with delivery address in request.');
+                cartValidationOutputCollection.add(cvo);
+				return;
+			}
+
+    		// Clean up CVO based on tax. When new tax calculator request comes, we need to clean up
+    		// previous CVOs as they have been previously handled by the Cart Calculate API.
+    		for (Integer i = (cartValidationOutputCollection.size() - 1); i >= 0; i--) {
+            	CartExtension.CartValidationOutput cvo = cartValidationOutputCollection.get(i);
+            	if (cvo.getType() == CartExtension.CartValidationOutputTypeEnum.TAXES) {
+            		cartValidationOutputCollection.remove(cvo);
+            	}
+            }
 
-      // There should be one delivery group per cart.
-      CartExtension.CartDeliveryGroupList cartDeliveryGroups = cart.getCartDeliveryGroups();
-      CartExtension.CartDeliveryGroup cartDeliveryGroup = cartDeliveryGroups.get(0);
+            // There should be one delivery group per cart.
```

Comment:
again - that is invalid assumption. Multiple delivery groups is a valid feature that SCom Cart and Checkout supports

Reply:
As discussed on slack this cls will consider only first delivery group and mentioned the same on start of apex class